### ISSUE TITLE
linuxloops hook

### ISSUE THREAD

**Issue by lavilao** [sebanc/linuxloops#8]
Hi (sorry its me again) does the linuxloops hook must remain in order for the grub config to grub? am trying to install (emphasis on trying) blend os 4 and it to update makes a new rootfs folder where it will install the desired packages ej: I installed arch with linuxloops (all good) but after installing akshara (basically like nix but for archlinux) it will create a new archlinux chroot (directly from the bootstrap image) inside .new_rootfs and after it finishes uses overlays (have not reached that point due to a bug) to replace the current root with .new_root. This new root does not include by default the linuxloops hook so it got me thinking if it is necesary to keep it for the grub inside the .img file works with the grub entry created by linuxloops in installation of arch. Also, is this enough to add the linuxloops hook?
```
exec('mkdir', '-p', '/.new_rootfs/usr/lib')
        exec('cp', '-r', '/usr/lib/linuxloops', '/.new_rootfs/usr/lib')
        exec('mkdir', '-p', '/.new_rootfs/etc/initcpio/install')
        exec('cp', '/etc/initcpio/install/linuxloops', '/.new_rootfs/etc/initcpio/install')
```
I basically just copy the hooks (that I found) from root to new_root at the start of the update and add the linuxloops hook to mkinitcpio of the new root with
```
exec_chroot('bash', '-c', 'echo "HOOKS=(base udev akshara autodetect keyboard keymap modconf block filesystems fsck linuxloops)" >> /etc/mkinitcpio.conf')
```
So, TL:DR is that all that is needed to keep the linuxloop hook working or is not needed after installation? Thanks in advice!
PD: adding the cps and linuxloops to mkinitcpio removes an error of linuxloops hook not found when update but I have not been able to see if it works because the update process is atomic and it fails because of a f*cking TypeError in python.

**Comment by lavilao** [comment]
Finally it installed, yes it works with the cp. Sorry for beign annoyimg

**Comment by sebanc** [comment]
Hi,

I can try to add BlendOS, I will need some time to look at the structure but I guess it should be possible.

For immutable distros, I use a different process by creating a secondary initramfs that contains the linuxloops binary and use a different grub config that loads both initramfs at boot time (BlissOS and Tails are examples).

I would be interested to know how you create the BlendOS rootfs from arch, it would probably save time.

**Comment by lavilao** [comment]
Blendos 4 is basically an Arch installation and update script. I just used the Arch installation of linuxloops to get Arch setup and then followed the instructions of the website. By default it gives error saying it could not find the linuxloops hook but it's just matter of copying then to the right locations in the new_rootfs at chroot time, I modified the script itself but it should not be necessary as the script (akshara) supports custom commands at chroot time so one should just add a command that downloads the hook files to the right locations and appends linuxloops to the hooks list inside the akshara of the new_rootfs and it should work.

**Comment by lavilao** [comment]
here is the guide: https://blog.asterisk.lol/blog/2023/11/blendos-v4-install-guide/#configuration-customization
and this is a example of how to download something at chroot time (taked from here: https://github.com/Sil3ntLight/blendos-tracks/blob/main/blendos-base.yaml):
```
commands:
  - 'curl https://raw.githubusercontent.com/Sil3ntLight/blendos-tracks/main/akshara -o /usr/bin/akshara && chmod 755 /usr/bin/akshara'
```

**Comment by lavilao** [comment]
sorry to ask you this out of nowhere  but are you able to update through OTA in chromeos flex installed to a .img file with linuxloops? I tried today and it never finished, the cpu was at 100% for about 1:30h, took ao long I had to shut down the laptop . wish there was some kind of progress bar or something, I had no way of knowing if the postinstall script had finished or not as the reboot to update notification never came. PD: the update was about 900mb in size.

**Comment by sebanc** [comment]
Thanks for the links !

I just tried ChromeOS-Flex, I updated from stable to beta and then from beta to dev without issue with each time the update taking 1 minute to complete after reaching 100% (I have an SSD). An initramfs with modules and firmwares is created when the update reach 100%, this process is ressource consuming but even for a hdd 1:30 is very long....

Also indeed if the update did not complete it should have booted the old rootfs.

Not sure what the issue is.

**Comment by lavilao** [comment]
> Thanks for the links !
> 
> I just tried ChromeOS-Flex, I updated from stable to beta and then from beta to dev without issue with each time the update taking 1 minute to complete after reaching 100% (I have an SSD). An initramfs with modules and firmwares is created when the update reach 100%, this process is ressource consuming but even for a hdd 1:30 is very long....
> 
> Also indeed if the update did not complete it should have booted the old rootfs.
> 
> Not sure what the issue is.
> 

Could it be because the partition it is in is ntfs? I Saw errors of flashrom but I thought it was for firmware, could it be that some operations could not be done in the .img? But then again You updated just fine. Bad luck I guess 🤷🏾‍♂️.

**Comment by sebanc** [comment]
Just a quick update to say I have added BlendOS v4 alpha support. So far, Plasma DE seems to be the most stable, the other ones have some bugs and Gnome won't boot on my laptop.

**Comment by lavilao** [comment]
Oh cool, sad to hear that gnome did not worked for You (it did for me). I reinstalled Flex, in dev mode this time, I have not updated yet. Installed Arch un crostini and everything seems to work fine, Will try to update tomorrow. One thing, I gave Flex 20gb, that should not matter right? I mean the a and b root spaces is already pre allocated right?

**Comment by sebanc** [comment]
Yes ChromeOS Flex should be able to update regardless of the image size.

**Comment by lavilao** [comment]
rebooted, same thing unbootable system (chromeos). tried to mount the img file in linux ```sudo losetup -fP --show flex.img```, the root-a partition could not be mounted. My theory is that on a ntfs drive chromeos updates fail due to partition corruption, need confirmation from other people thou.

**Comment by sebanc** [comment]
> trying to update again, stuck at postinstall [update_engine.log](https://github.com/sebanc/linuxloops/files/13964195/update_engine.log) do you see anything out of place in the logs?

I just tried ChromeOS-Flex on a ntfs partition and updated from stable to beta and then from beta to dev without issue.
The log you posted also does not show any specific issue.
At this stage my only guess so far is that you are using a very slow drive.

> rebooted, same thing unbootable system (chromeos). tried to mount the img file in linux `sudo losetup -fP --show flex.img`, the root-a partition could not be mounted. My theory is that on a ntfs drive chromeos updates fail due to partition corruption, need confirmation from other people thou.

ChromeOS rootfs partitions are protected and can only be mounted read-only, it will fail if you do not specify "-o ro" as an arugment of the mount command.

**Comment by lavilao** [comment]
the only other thing I could find were errors like this one ```ERR kernel: [   21.247939] blk_update_request: operation not supported error, dev loop0, sector 27324456 op 0x3:(DISCARD) flags 0x800 phys_seg 1 prio class 0```, this is from messages.log. I just dont undestand why it does not boot! does the windows version affects anything? am using win 10 1903.

**Comment by lavilao** [comment]
[messages.log](https://github.com/sebanc/linuxloops/files/13982157/messages.log)

**Comment by sebanc** [comment]
> the only other thing I could find were errors like this one `ERR kernel: [ 21.247939] blk_update_request: operation not supported error, dev loop0, sector 27324456 op 0x3:(DISCARD) flags 0x800 phys_seg 1 prio class 0`, this is from messages.log. I just dont undestand why it does not boot! does the windows version affects anything? am using win 10 1903.

The above errors are harmless and the log shows nothing significant (there are still a lot of curl related errors that I am not used to see in the logs but I don't think it is related).
Might your ntfs partition be somehow corrupt ?

**Comment by lavilao** [comment]
> > the only other thing I could find were errors like this one `ERR kernel: [ 21.247939] blk_update_request: operation not supported error, dev loop0, sector 27324456 op 0x3:(DISCARD) flags 0x800 phys_seg 1 prio class 0`, this is from messages.log. I just dont undestand why it does not boot! does the windows version affects anything? am using win 10 1903.
> 
> The above errors are harmless and the log shows nothing significant (there are still a lot of curl related errors that I am not used to see in the logs but I don't think it is related).
> Might your ntfs partition be somehow corrupt ?

My ntfs partition should not be corrupted, it's the c drive so Windows would immediately complain about it. However thats a "should", I do remember that I had to stop using brunch cuz it did not booted with a reboot saying filesystem error, after entering to Windows the partition was marked as dirty and then I did a chkdsk and the .img file got deleted, but with chromeos Flex the system works fine until I update. And it's not like it works because I don't touch anything cuz I installed crostini, Google assistant and other DLC. Will have to check puting the img file on the d partition of Windows to check.

**Comment by SROG33** [comment]
My ChromeOS Flex install is also unbootable after I tried to update to beta channel. The download was stucked at 100% so I tried to restart but It didn't boot. I installed it on a separate ext4 partition with only the image file. I think it has something to do with the Linux subsystem because before using Linux I updated from 119 to 120 and it worked fine.

**Comment by lavilao** [comment]
> Thanks to you both for your feedbacks.
> 
> I was eventually able to reproduce the bug and pushed a potential fix.
> 
> Could you try a fresh install ?

testing now, will try to update after setting linux up with my container to try the fix in a real world scenario

**Comment by lavilao** [comment]
It worked! I was able to update! updated from 120.0.6099.203 to 120.0.6099.235, with linux enviroment installed and all. Confused at why such little update would require a full payload (900+ MB) but minor details, thanks @sebanc for beign so patient.

**Comment by SROG33** [comment]
Worked for me as well! Thanks @sebanc.

**Comment by sebanc** [comment]
Thanks for the feedbacks !

Actually, delta updates are disabled when booting ChromeOS-Flex from an image as I have to modify 2 files on the main roots to boot from the image and support updates. Unfortunately, there is not much I can do about it

**Comment by lavilao** [comment]
> Thanks for the feedbacks !
> 
> Actually, delta updates are disabled when booting ChromeOS-Flex from an image as I have to modify 2 files on the main roots to boot from the image and support updates. Unfortunately, there is not much I can do about it 

Oh, that explains why logs said that ibas using a dev/test image
